search/includes/classes/class-queue.php

```diff
@@ -0,0 +1,350 @@
+	}
+
+	public function update_jobs( $job_ids, $data ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		$escaped_fields = [];
+
+		foreach ( $data as $column => $value ) {
+			$escaped_fields[] = $wpdb->prepare( "{$column} = %s", $value ); // Cannot prepare column name. @codingStandardsIgnoreLine
+		}
+
+		$escaped_fields = implode( ', ', $escaped_fields );
+
+		$escaped_ids = implode( ', ', array_map( 'intval', $job_ids ) );
+
+		$sql = "UPDATE {$table_name} SET {$escaped_fields} WHERE id IN ( {$escaped_ids} )"; // Cannot prepare table name. @codingStandardsIgnoreLine
+
+		return $wpdb->get_results( $sql ); // Already escaped. @codingStandardsIgnoreLine
+	}
+
+	public function delete_jobs( $jobs ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		$ids = wp_list_pluck( $jobs, 'id' );
+
+		$escaped_ids = implode( ', ', array_map( 'intval', $ids ) );
+
+		$sql = "DELETE FROM {$table_name} WHERE id IN ( {$escaped_ids} )"; // Cannot prepare table name. @codingStandardsIgnoreLine
+
+		return $wpdb->get_results( $sql ); // Already escaped. @codingStandardsIgnoreLine
+	}
+
+	public function empty_queue() {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		return $wpdb->query( "TRUNCATE TABLE {$table_name}" ); // Cannot prepare table name. @codingStandardsIgnoreLine
+	}
+
+	public function count_jobs( $status, $object_type = 'post' ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT COUNT(*) FROM {$table_name} WHERE `status` = %s AND `object_type` = %s", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$status,
+				$object_type
+			)
+		);
+	}
+
+	public function count_jobs_due_now( $object_type = 'post' ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		return $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT COUNT(*) FROM {$table_name} WHERE `status` = 'queued' AND `object_type` = %s AND ( `start_time` <= NOW() OR `start_time` IS NULL )", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$object_type
+			)
+		);
+	}
+
+	public function get_next_job_for_object( $object_id, $object_type ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		$job = $wpdb->get_row(
+			$wpdb->prepare(
+				"SELECT * FROM {$table_name} WHERE `object_id` = %d AND `object_type` = %s AND `status` = 'queued' LIMIT 1", // Cannot prepare table name. @codingStandardsIgnoreLine
+				$object_id,
+				$object_type
+			)
+		);
+
+		return $job;
+	}
+
+	public function get_batch_jobs( $count = 250 ) {
+		global $wpdb;
+
+		$table_name = $this->schema->get_table_name();
+
+		// TODO transaction
+
+		$jobs = $wpdb->get_results(
+			$wpdb->prepare(
+				"SELECT * FROM {$table_name} WHERE ( `start_time` <= NOW() OR `start_time` IS NULL ) AND `status` = 'queued' LIMIT %d", // Cannot prepare table name. @codingStandardsIgnoreLine
```

Comment:
Do we need to factor in priority here?

Reply:
Definitely, but that'll be a "phase 2" improvement...for now I'm prioritizing just deduplicating bulk re-index requests which is the immediate need.